---
title: "Using the scREhurdle package"
author: "Michael Sekula"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scREhurdle}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette demonstrates how to use the `scREhurdle` package to perform a differential expression analysis on discrete single-cell RNA-seq (scRNA-seq) data. It should be noted that `scREhurdle` imports the `rstan` package to perform approximate Bayesian inference using automatic differentiation variational inference. Details about `rstan` can be found [here](https://mc-stan.org/users/interfaces/rstan).

```{r, warning=FALSE}
# Load the scREhurdle package
library(scREhurdle)
```

## The Data

`scREhurdle` requires discrete scRNA-seq data, so either a matrix or data frame of read counts is acceptable. The rows of the data correspond to genes and the columns correspond to cells. For the purpose of this vignette, it is assumed that the dataset has already been cleaned and is ready to analyze. An example dataset is included with the `scREhurdle` package.

```{r}
# Load data
data(toyDat)
scDat <- toyDat$scDat
head(scDat)
```

The `toyDat` object is actually a list containing two separate objects: `scDat` and `subpop`. The example dataset is `scDat`, which consists of 100 rows (genes) and 50 columns (cells). In this data, the first 25 cells belong to the control group (C) and the remaining 25 cells belong to the treatment group (T).

### Subpopulation Clustering (Recommended but Optional)
The second object `subpop` contains the subpopulation assignment for each cell. For a real data analysis, these assignments can be determined by a clustering algorithm suitable for clustering scRNA-seq data (e.g., SNN-Cliq, SC3, etc.). The clusters should be determined separately for each treatment group. This means there should be no overlap in the cluster assignments between cells in different treatment groups. From the output below we see the first 25 cells (in the control group) are clustered into two groups (1 and 2), while the next 25 cells (in the treatment group) are clustered into three groups (3, 4, and 5).

```{r}
# Extract example subpopulation clustering assignment
subpop <- toyDat$subpop
subpop[1:25]
subpop[26:50]
```

Including a subpopulation assignment is recommended for data analysis, but it is not required to run the `scREhurdle` function. 

## Differential Expression Analysis

The `scREhurdle` function uses Stan's variational algorithm for approximate posterior sampling to determine differentially expressed (DE) genes. Essentially, `scREhurdle` is a wrapper function that allows users to interface with the `vb` function in `rstan` when implementing the hurdle model methodology proposed in the "Detection of differentially expressed genes in discrete single-cell RNA sequencing data using a hurdle model with correlated random effects" manuscript. Details about the `vb` function can be found at the [RStan website](https://mc-stan.org/users/interfaces/rstan).

The main parameters of the `scREhurdle` function are listed below:

* `Y` - matrix or data.frame where the rows correspond to genes and columns correspond to cells.
* `treatGroup` - vector or factor indicating the treatment assignments (control or treatment) of the cells.
* `useCDR` - whether CDR, the proportion of genes expressed per cell, is calculated and used in model matrix.
* `typeRE` - set the type of random effects to use.
* `subpop` - vector or factor indicating the subpopulation assignment of cells (used ONLY in the correlated random effects model).

These parameters help the user define the modeling structure for differential expression analysis. All parameters of the `vb` function can be passed through `scREhurdle`.

### Correlated Random Effects (CRE) Model

The CRE model requires the use of a subpopulation structure determined by a cluster analysis of the cells. *See the next section on Independent Random Effects if no subpopulation structure is being used.* 

To analyze discrete scRNA-seq data with the CRE model, the following input is required:

* `Y` - count (integer) data as a matrix or data.frame.
* `treatGroup` - vector or factor with the treatment assignments (control or treatment) of the cells.
* `typeRE` - set to "cor".
* `subpop` - vector or factor indicating the subpopulation assignment of cells.

As default, the proportion of genes expressed per cell is used as a covariate in the model matrix. While it is recommended to use this covariate, adding `useCDR = FALSE` as an argument to the `scREhurdle` function will exclude it from the analysis. 

A seed for the `vb` function is also included in the example code below to ensure reproducible results.

```{r, results='hide'}
# Obtain treatment assignments
treat <- substr(colnames(scDat), start = 1, stop = 1)

# Analysis with CRE model
mod.CRE <- scREhurdle(Y=scDat, treatGroup=treat, typeRE="cor", subpop=subpop, stan_seed=523)

# For the sake of brevity, the following is example output
# produced when running the scREhurdle function
#>
#> Running CRE Stan model... 
#> Chain 1: ------------------------------------------------------------
#> Chain 1: EXPERIMENTAL ALGORITHM:
#> Chain 1:   This procedure has not been thoroughly tested and may be unstable
#> Chain 1:   or buggy. The interface is subject to change.
#> Chain 1: ------------------------------------------------------------
#> Chain 1: 
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Gradient evaluation took 0.00357 seconds
#> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 35.7 seconds.
#> Chain 1: Adjust your expectations accordingly!
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Begin stochastic gradient ascent.
#> Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes 
#> Chain 1:    100       -18252.944             1.000            1.000
#> Chain 1:    200       -16177.294             0.564            1.000
#> Chain 1:    300       -15904.037             0.382            0.128
#> Chain 1:    400       -15781.913             0.288            0.128
#>  ...       ...               ...              ...              ...
#> Chain 1:   7800       -14849.884             0.000            0.000
#> Chain 1:   7900       -14849.125             0.000            0.000   MEAN ELBO CONVERGED
#> Chain 1: 
#> Chain 1: Drawing a sample of size 1000 from the approximate posterior... 
#> Chain 1: COMPLETED.
```

After computation, `mod.CRE` is an object of class `scREhurdle.fit`. Information relevant to differential expression (i.e., parameter estimates and test statistics of the treatment indicator variables) can be extracted with `$deTab`.

```{r}
## Table with DE analysis results
DEtab1 <- mod.CRE$deTab
head(DEtab1)
```

Parameter estimates, Z test statistics, and unadjusted p-values are displayed in the first six columns of `DEtab1`. Since a hurdle model has been fit, there are two sets of parameters and associated test statistics: one set from the logistic regression and one set from zero-truncated negative binomial regression. The first three columns represent the values associated with the logistic regression (note the "L" in the column names). The next three columns represent the values associated with count model (note the "C" in the column names). While this information is useful in determining whether genes are DE in terms of proportion of zeros or actual expression, more often the interest is in a single combined result. The `chisq` column provides the two-dimensional test statistics that combine information from both parts of the hurdle model to determine DE genes. The final two columns present the unadjusted and adjusted p-values of these test statistics. By default, p-values are adjusted with the Benjamini-Hochberg method.

```{r}
# Obtain gene names
gene.names <- rownames(scDat)

# DE genes at 5% FDR
de.genes <- which(DEtab1$chisq.padj <= 0.05)

length(de.genes)
gene.names[de.genes]

```

Using the adjusted p-values in the last column of `DEtab1`, there are 48 genes significant at a 5% FDR.

### Independent Random Effects (IRE) Model

Unlike the CRE model, the IRE model does not require the use of a subpopulation structure determined by a cluster analysis of the cells. Under this model, the random effects are assumed to be independent of one another. The example provided in the previous section on the Correlated Random Effects model is repeated in this section. The only difference here is that the Independent Random Effects model is being used. 

To analyze discrete scRNA-seq data with the IRE model, the following input is required:

* `Y` - count (integer) data as a matrix or data.frame.
* `treatGroup` - vector or factor with the treatment assignments (control or treatment) of the cells.
* `typeRE` - set to "ind".

There is no need to use the `subpop` parameter when the random effects are treated as independent.

As default, the proportion of genes expressed per cell is used as a covariate in the model matrix. While it is recommended to use this covariate, adding `useCDR = FALSE` as an argument to the `scREhurdle` function will exclude it from the analysis. 

A seed for the `vb` function is also included in the example code below to ensure reproducible results.

```{r, results='hide'}
# Obtain treatment assignments
treat <- substr(colnames(scDat), start = 1, stop = 1)

# Analysis with IRE model
mod.IRE <- scREhurdle(Y=scDat, treatGroup=treat, typeRE = "ind", stan_seed=523)

# For the sake of brevity, the following is example output
# produced when running scREhurdle
#> 
#> Running IRE Stan model... 
#> Chain 1: ------------------------------------------------------------
#> Chain 1: EXPERIMENTAL ALGORITHM:
#> Chain 1:   This procedure has not been thoroughly tested and may be unstable
#> Chain 1:   or buggy. The interface is subject to change.
#> Chain 1: ------------------------------------------------------------
#> Chain 1: 
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Gradient evaluation took 0.003562 seconds
#> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 35.62 seconds.
#> Chain 1: Adjust your expectations accordingly!
#> Chain 1: 
#> Chain 1: 
#> Chain 1: Begin stochastic gradient ascent.
#> Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes 
#> Chain 1:    100       -18172.718             1.000            1.000
#> Chain 1:    200       -15942.644             0.570            1.000
#> Chain 1:    300       -15743.173             0.384            0.140
#> Chain 1:    400       -15636.951             0.290            0.140
#>    ...     ...               ...              ...              ...
#> Chain 1:   5100       -14837.830             0.000            0.000
#> Chain 1:   5200       -14837.006             0.000            0.000   MEDIAN ELBO CONVERGED
#> Chain 1: 
#> Chain 1: Drawing a sample of size 1000 from the approximate posterior... 
#> Chain 1: COMPLETED.
```

After computation, `mod.IRE` is an object of class `scREhurdle.fit`. Information relevant to differential expression (i.e., parameter estimates and test statistics of the treatment indicator variables) can be extracted with `$deTab`.

```{r}
## Table with DE analysis results
DEtab2 <- mod.IRE$deTab
head(DEtab2)
```

Parameter estimates, Z test statistics, and unadjusted p-values are displayed in the first six columns of `DEtab2`. Since a hurdle model has been fit, there are two sets of parameters and associated test statistics: one set from the logistic regression and one set from zero-truncated negative binomial regression. The first three columns represent the values associated with the logistic regression (note the "L" in the column names). The next three columns represent the values associated with count model (note the "C" in the column names). While this information is useful in determining whether genes are DE in terms of proportion of zeros or actual expression, more often the interest is in a single combined result. The `chisq` column provides the two-dimensional test statistics that combine information from both parts of the hurdle model to determine DE genes. The final two columns present the unadjusted and adjusted p-values of these test statistics. By default, p-values are adjusted with the Benjamini-Hochberg method.

```{r}
# Obtain gene names
gene.names <- rownames(scDat)

# DE genes at 5% FDR
de.genes <- which(DEtab2$chisq.padj <= 0.05)

length(de.genes)
gene.names[de.genes]

```

Using the adjusted p-values in the last column of `DEtab2`, there are 49 genes significant at a 5% FDR.
